### PR TITLE
REST API Layer changes for recommission support

### PR DESCRIPTION
**Pull request by pranikum**
Signed-off-by: pranikum <109206473+pranikum@users.noreply.github.com>

### Description
This change contain REST API layer changes for Recommission API support. 
This pr is dependent of Service layer changes. (https://github.com/opensearch-project/OpenSearch/pull/4320)

This change has some changes for the PR https://github.com/opensearch-project/OpenSearch/pull/4084

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by imRishN** [comment]
Changes to - 
1. `client/rest-high-level/src/test/java/org/opensearch/client/RestHighLevelClientTests.java`
2. `rest-api-spec/src/main/resources/rest-api-spec/api`
are missing. We need to add that

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionRequest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+    }
+
+    @Override
+    public String toString() {
+        return "DeleteDecommissionRequest";
+    }
+}
```
[the remark this answers is not in the document]
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionRequest.java]: Removed.

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionAction.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    protected void masterOperation(
+        DeleteDecommissionRequest request,
+        ClusterState state,
+        ActionListener<DeleteDecommissionResponse> listener
+    ) {
+        // TODO: Enable when service class change is merged
+        logger.info("Received delete decommission Request");
+        listener.onResponse(new DeleteDecommissionResponse(true));
+    }
+}
```
[the remark this answers is not in the document]
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionAction.java]: Done.

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java [hunk 3]
```diff
@@ -791,4 +794,19 @@ public interface ClusterAdminClient extends OpenSearchClient {
      * Delete specified dangling indices.
      */
     ActionFuture<AcknowledgedResponse> deleteDanglingIndex(DeleteDanglingIndexRequest request);
+
+    /**
+     * Decommission a node
+     */
```
[the remark this answers is not in the document]
Comment (pranikum) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java]: Updated.

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionAttribute.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.decommission;
```
[the remark this answers is not in the document]
Comment (pranikum) [reply server/src/main/java/org/opensearch/cluster/decommission/DecommissionAttribute.java]: Done.

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionAction.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.opensearch.action.ActionType;
+
+/**
+ * Transport action delete decommission.
+ *
+ * @opensearch.internal
+ */
+
+public class DeleteDecommissionAction extends ActionType<DeleteDecommissionResponse> {
```
Comment (imRishN) [inline server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionAction.java:19]: As part of one comment of this PR - https://github.com/opensearch-project/OpenSearch/pull/4261, we decided to rename it to DeleteDecommissionStateAction

Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionAction.java]: Done.

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java [hunk 6]
```diff
@@ -791,4 +794,19 @@ public interface ClusterAdminClient extends OpenSearchClient {
...
+
+    /**
+     * Deletes the decommission metadata.
+     */
+    void deleteDecommission(DeleteDecommissionStateRequest request, ActionListener<DeleteDecommissionStateResponse> listener);
+
+    /**
+     * Deletes the decommission metadata.
+     */
+    DeleteDecommissionStateRequestBuilder prepareDeleteDecommission();
```
Comment (imRishN) [inline server/src/main/java/org/opensearch/client/ClusterAdminClient.java:811]: nit: deleteDecommissionState

File: server/src/main/java/org/opensearch/client/Requests.java [hunk 7]
```diff
@@ -539,6 +540,13 @@ public static DeleteSnapshotRequest deleteSnapshotRequest(String repository, Str
         return new DeleteSnapshotRequest(repository, snapshots);
     }
 
+    /**
+     * Creates a new delete decommission request.
+     */
+    public static DeleteDecommissionStateRequest deleteDecommissionRequest() {
```
Comment (imRishN) [inline server/src/main/java/org/opensearch/client/Requests.java:546]: nit: deleteDecommissionStateRequest

**Review by imRishN — APPROVED** [review]

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java (hunk 6 above)
Comment (pranikum) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java]: Done.

File: server/src/main/java/org/opensearch/client/Requests.java (hunk 7 above)
Comment (pranikum) [reply server/src/main/java/org/opensearch/client/Requests.java]: Done.

**Comment by pranikum** [comment]
Closing this PR. Reopened another one.

https://github.com/opensearch-project/OpenSearch/pull/4604

**Pull request closed without merging**